### PR TITLE
Webhook does not process transient errors properly

### LINKED ISSUE #2173: Webhook does not process transient errors properly
### MetalLB Version

latest

### Deployment method

Not relevant

### Main CNI

N/A

### Kubernetes Version

_No response_

### Cluster Distribution

_No response_

### Describe the bug

Using the webhooks, we do not want to block resources depending on the order they are received - e.g we don't want to block a peer specifying a bfdProfile that doesn't exist yet because it could eventually converge. today we achieve that using the TransientError  type, returning it only when an error that relates to dependencies between resources occur and we ignore this type of err in the webhook's validator.
The problem with that approach is that in some cases this causes the webhook to miss an actual error, because in the webhook we pass multiple objects of the same type to the validation and return early when hitting any error, which for example if a TransientError happens for the first resource the rest of the resources are not being validated at all.
As an example, consider the following resources:
```yaml
apiVersion: metallb.io/v1beta2
kind: BGPPeer
metadata:
  name: peer1
  namespace: metallb-system
spec:
  peerAddress: 172.16.30.2
  peerASN: 65433
  myASN: 65439
  bfdProfile: "non-existing"
---
apiVersion: metallb.io/v1beta2
kind: BGPPeer
metadata:
  name: peer2
  namespace: metallb-system
spec:
  peerAddress: 172.16.30.3
  peerASN: 0
  myASN: 65439
```
peer2 is invalid since it specifies peerASN: 0  so we would expect the webhook to block it anyways, but the outcome of the webhook changes as a result of the order: if peer2 is created without peer1 it is blocked, if it is created after peer1 it passes.
A possible solution is instead of ignoring transient errors in the webhook's validator, it would pass the resources to `For` without fields that can cause a transient error (resetting them), essentially not allowing the webhook to receive a transient error at all.
TransientErrors could stay in the codebase (serving as a "doc"), we would just not act on them in the webhooks.

### To Reproduce

described above

### Expected Behavior

described above

### Additional Context

described above

### I've read and agree with the following

- [X] I've checked all open and closed issues and my request is not there.
- [X] I've checked all open and closed pull requests and my request is not there.

### I've read and agree with the following

- [X] I've checked all open and closed issues and my issue is not there.
- [X] This bug is reproducible when deploying MetalLB from the main branch
- [X] I have read the [troubleshooting guide](https://metallb.universe.tf/troubleshooting/) and I am still not able to make it work
- [X] I checked the logs and MetalLB is not discarding the configuration as not valid
- [X] I enabled the debug logs, collected the information required from the cluster using the [collect script](https://raw.githubusercontent.com/metallb/metallb/main/troubleshooting/collect.sh) and will attach them to the issue
- [X] I will provide the definition of my service and the related endpoint slices and attach them to this issue

### PR DESCRIPTION
**Pull request by shimritproj**
Fixed  #2173 

There are TransientErrors that can occasionally result in overlooking genuine issues. The solution is to reset the fields that might be causing a TransientError.

### CODE REVIEW

File: internal/config/validator.go [hunk 1]
```diff
@@ -44,14 +44,35 @@ func (v *validator) Validate(resources ...client.ObjectList) error {
...
+			for i := range (clusterResources.LegacyAddressPools) {
+				for j := range(clusterResources.LegacyAddressPools[i].Spec.BGPAdvertisements) {
+					for k := range(clusterResources.LegacyAddressPools[i].Spec.BGPAdvertisements[j].Communities){
+						if !existString(clusterResources.LegacyAddressPools[i].Spec.BGPAdvertisements[j].Communities[k], clusterResources.Communities) {
+							clusterResources.LegacyAddressPools[i].Spec.BGPAdvertisements[j].Communities = removeElement(clusterResources.LegacyAddressPools[i].Spec.BGPAdvertisements[j].Communities,
+								 clusterResources.LegacyAddressPools[i].Spec.BGPAdvertisements[j].Communities[k])
+						}
+					}
+				}
+			}
```
Comment (liornoy) [inline internal/config/validator.go:75]: Using indexes to reference slice items make it very hard to read.
You can the second returned value (the object itself) from range as following:
```
for _, legacyAddrPool := range clusterResources.LegacyAddressPools {
    for _, bgpAdv := range legacyAddrPool {
    ...
    
```

File: internal/config/validator.go [hunk 2]
```diff
@@ -68,3 +89,32 @@ func (v *validator) Validate(resources ...client.ObjectList) error {
 func NewValidator(validate Validate) apivalidate.ClusterObjects {
 	return &validator{validate: validate}
 }
+
+func existString(c string, community []metallbv1beta1.Community ) bool {
```
[the remark this answers is not in the document]
Comment (liornoy) [reply internal/config/validator.go:93]: can we please also rename the function because "existString" is too generic, and that's specifically to check if a community exists

**Review by liornoy — COMMENTED** [review]
Thanks for the PR!
Left some comments, and please write a commit message :)

File: internal/config/validator.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (shimritproj) [reply internal/config/validator.go:93]: I changed the name of the function and the arguments to be like this :  `func ExistCommunity(community string, communities []metallbv1beta1.Community )`

File: internal/config/validator.go (hunk 1 above)
Comment (shimritproj) [reply internal/config/validator.go:75]: I agree and changed it.

File: internal/config/validator.go [hunk 3]
```diff
@@ -44,14 +44,35 @@ func (v *validator) Validate(resources ...client.ObjectList) error {
 			clusterResources.Pools = append(clusterResources.Pools, list.Items...)
```
[the remark this answers is not in the document]
Comment (shimritproj) [reply internal/config/validator.go:44]: I disagree because I don't understand why it is better than the existing one.

[the remark this answers is not in the document]
Comment (liornoy) [reply internal/config/validator.go:44]: I think it would be better to move the override logic to a function because this will make the switch-case block less cluttered so it is easier to read and understand.
We do extra logic for peers, bgpadv, and addresspool, and it might confuse the reader.
With a function the code will look cleaner, and we can add a small comment explaining why we do this non-trivial override.

[the remark this answers is not in the document]
Comment (shimritproj) [reply internal/config/validator.go]: I agree that an additional function should be created whose role will only be to reset the fields that cause TransientError.

File: internal/config/validator.go [hunk 4]
```diff
@@ -58,13 +58,69 @@ func (v *validator) Validate(resources ...client.ObjectList) error {
 			clusterResources.Nodes = append(clusterResources.Nodes, list.Items...)
 		}
 	}
+	clusterResources = Resetfields(clusterResources)
```
Comment (liornoy) [inline internal/config/validator.go:61]: Let's keep this function private to this package

**Comment by liornoy** [comment]
Sorry for not saying this before, but now that I think of this, we also don't need to export the other 2 functions - `RemoveElement` and `ExistCommunity`. 
And can you please run "go fmt" on it? this has some formatting issues.

**Comment by shimritproj** [comment]
> Sorry for not saying this before, but now that I think of this, we also don't need to export the other 2 functions - `RemoveElement` and `ExistCommunity`. And can you please run "go fmt" on it? this has some formatting issues.

Yes, sure! you can see now that the file is updated.

File: internal/config/validator.go [hunk 5]
```diff
@@ -58,13 +58,69 @@ func (v *validator) Validate(resources ...client.ObjectList) error {
...
 	}
+	clusterResources = resetfields(clusterResources)
 	_, err := For(clusterResources, v.validate)
 	if errors.As(err, &TransientError{}) { // we do not want to make assumption on ordering in webhooks.
 		return nil
 	}
 	return err
 }
 
+func resetfields(clusterResources ClusterResources) ClusterResources {
```
Comment (fedepaol) [inline internal/config/validator.go:69]: resetfields is a bit too generic. Please add a proper name and a comment that says what are the fields that we are resetting.

File: internal/config/validator.go [hunk 6]
```diff
@@ -58,13 +58,69 @@ func (v *validator) Validate(resources ...client.ObjectList) error {
...
+	for i := range clusterResources.Peers {
+		clusterResources.Peers[i].Spec.BFDProfile = ""
+		clusterResources.Peers[i].Spec.PasswordSecret.Name = ""
+	}
+	for _, bgpAdv := range clusterResources.BGPAdvs {
+		for _, community := range bgpAdv.Spec.Communities {
+			if !ExistCommunity(community, clusterResources.Communities) {
+				bgpAdv.Spec.Communities = RemoveElement(bgpAdv.Spec.Communities, community)
```
Comment (fedepaol) [inline internal/config/validator.go:77]: I am not sure about what happens if we remove an item / reassing while we iterate.
A better approach is to build the new communities slice while iterate, and replace bgpAdv.Spec.Communities with the new slice after the loop

**Comment by fedepaol** [comment]
Do we need transient error after these changes?

**Comment by fedepaol** [comment]
Also, please change the commit message explaining the why we are doing this change.

File: internal/config/validator.go (hunk 5 above)
Comment (shimritproj) [reply internal/config/validator.go]: I agree , I changed the name to be `resetTransientErrorsFields`.

**Comment by shimritproj** [comment]
> Do we need transient error after these changes?
> 
> And, can you add unit tests?

On the Webhooks side, resetting the fields that cause Transienterrors, and there will be no Transienterrors after the changes. Transienterrors prevent us from seeing real errors. However, on the Speaker side, the Transienterrors will remain as usual. This is because, in the Speaker, the Transienterrors help us document the code. 
As for unit tests - yes, I will do them.
@fedepaol

**Comment by fedepaol** [comment]
> > Do we need transient error after these changes?
> > And, can you add unit tests?
> 
> On the Webhooks side, resetting the fields that cause Transienterrors, and there will be no Transienterrors after the changes. Transienterrors prevent us from seeing real errors. However, on the Speaker side, the Transienterrors will remain as usual. This is because, in the Speaker, the Transienterrors help us document the code.> Do we need transient error after these changes?
> And, can you add unit tests?

> However, on the Speaker side, the Transienterrors will remain as usual. This is because, in the Speaker, the 
> Transienterrors help us document the code. As for unit tests - yes, I will do them. @fedepaol

Not sure I understood this. Can you expand a bit?

> As for unit tests - yes, I will do them. @fedepaol

**Comment by liornoy** [comment]
> > > Do we need transient error after these changes?
> > > And, can you add unit tests?
> > 
> > 
> > On the Webhooks side, resetting the fields that cause Transienterrors, and there will be no Transienterrors after the changes. Transienterrors prevent us from seeing real errors. However, on the Speaker side, the Transienterrors will remain as usual. This is because, in the Speaker, the Transienterrors help us document the code.> Do we need transient error after these changes?
> > And, can you add unit tests?
> 
> > However, on the Speaker side, the Transienterrors will remain as usual. This is because, in the Speaker, the
> > Transienterrors help us document the code. As for unit tests - yes, I will do them. @fedepaol
> 
> Not sure I understood this. Can you expand a bit?
> 
> > As for unit tests - yes, I will do them. @fedepaol

Regarding transient errors, it might make sense to maintain the transient error as a custom error type. We can then return this error type in certain functions in config.go. This approach will make it clearer why we override some fields in the webhook.

File: internal/config/validator.go [hunk 7]
```diff
@@ -58,13 +56,51 @@ func (v *validator) Validate(resources ...client.ObjectList) error {
...
+		clusterResources.Peers[i].Spec.PasswordSecret.Name = ""
+	}
+	for i, bgpAdv := range clusterResources.BGPAdvs {
+		var communities []string
+		for _, community := range bgpAdv.Spec.Communities {
+			if existCommunity(community, clusterResources.Communities) {
+				communities = append(communities, community)
+			}
+		}
+		clusterResources.BGPAdvs[i].Spec.Communities = communities
```
Comment (liornoy) [inline internal/config/validator.go:77]: now that you're using the second value of the range you don't need to use indexes,
same for all loops here

File: internal/config/validator.go [hunk 8]
```diff
@@ -58,13 +56,51 @@ func (v *validator) Validate(resources ...client.ObjectList) error {
...
+// resetting the fields that lead to Transienterror
+func resetTransientErrorsFields(clusterResources ClusterResources) ClusterResources {
+	for i := range clusterResources.Peers {
+		clusterResources.Peers[i].Spec.BFDProfile = ""
+		clusterResources.Peers[i].Spec.PasswordSecret.Name = ""
+	}
+	for i, bgpAdv := range clusterResources.BGPAdvs {
+		var communities []string
+		for _, community := range bgpAdv.Spec.Communities {
+			if existCommunity(community, clusterResources.Communities) {
```
Comment (liornoy) [inline internal/config/validator.go:73]: wouldn't this make us miss real failures like a community with an invalid value?
like "234.4"

File: internal/config/validator.go (hunk 1 above)
Comment (liornoy) [reply internal/config/validator.go]: we still use indexes in existCommunity and resetTransientErrorsFields

**Comment by liornoy** [comment]
Can you please place "resetTransientErrorsFields" below "NewValidator" ?
it's a common practice to place the exported functions first and the private one after.

**Comment by shimritproj** [comment]
> > > > Do we need transient error after these changes?
> > > > And, can you add unit tests?
> > > 
> > > 
> > > On the Webhooks side, resetting the fields that cause Transienterrors, and there will be no Transienterrors after the changes. Transienterrors prevent us from seeing real errors. However, on the Speaker side, the Transienterrors will remain as usual. This is because, in the Speaker, the Transienterrors help us document the code.> Do we need transient error after these changes?
> > > And, can you add unit tests?
> > 
> > 
> > > However, on the Speaker side, the Transienterrors will remain as usual. This is because, in the Speaker, the
> > > Transienterrors help us document the code. As for unit tests - yes, I will do them. @fedepaol
> > 
> > 
> > Not sure I understood this. Can you expand a bit?
> > > As for unit tests - yes, I will do them. @fedepaol
> 
> Regarding transient errors, it might make sense to maintain the transient error as a custom error type. We can then return this error type in certain functions in config.go. This approach will make it clearer why we override some fields in the webhook.

A transient error was preventing me from identifying real errors, so we decided to remove them. If it is crucial for us to retain the information contained in errors, we can proceed with Lior's suggestion. 

@liornoy  @fedepaol

File: internal/config/validator.go (hunk 7 above)
Comment (liornoy) [reply internal/config/validator.go:77]: sorry, my mistake, using this value wont change the real slice, you are correct

File: internal/config/validator.go (hunk 8 above)
Comment (shimritproj) [reply internal/config/validator.go:77]: In my opinion, no, because the verification of whether the values are correct or not should occur in the preceding functions. In the functions here, I only delete or change to an empty string. 
@liornoy

Comment (shimritproj) [reply internal/config/validator.go:77]: After discussing with Lior, I agree.
Thanks @liornoy  for your attention here.

File: internal/config/validator.go [hunk 9]
```diff
@@ -58,13 +58,65 @@ func (v *validator) Validate(resources ...client.ObjectList) error {
...
+		bgpAdv.Spec.Communities = communities
+	}
+	for i, legacyAddrPool := range clusterResources.LegacyAddressPools {
+		for j, bgpAdv := range legacyAddrPool.Spec.BGPAdvertisements {
+			var communities []string
+			for k := range bgpAdv.Communities {
+				if existCommunity(bgpAdv.Communities[k], clusterResources.Communities) {
+					communities = append(communities, bgpAdv.Communities[k])
+				} else if !checkInvalidCommunityFormat(bgpAdv.Communities[k]) {
+					communities = append(communities, bgpAdv.Communities[k])
```
Comment (liornoy) [inline internal/config/validator.go:95]: Thanks, can you please include this corner case in the new unit test you wrote?

File: internal/config/validator.go [hunk 10]
```diff
@@ -58,13 +58,65 @@ func (v *validator) Validate(resources ...client.ObjectList) error {
...
+		clusterResources.Peers[i].Spec.PasswordSecret.Name = ""
+	}
+	for i, bgpAdv := range clusterResources.BGPAdvs {
+		var communities []string
+		for _, community := range bgpAdv.Spec.Communities {
+			if existCommunity(community, clusterResources.Communities) {
+				communities = append(communities, community)
+			} else if !checkInvalidCommunityFormat(community) {
+				communities = append(communities, community)
+			}
```
Comment (oribon) [inline internal/config/validator.go:83]: I think this gets into the validation details too much, rather than focusing on "avoiding transient errors". What I mean is what we actually want is to not pass references to other objects (like with the bfdprofiles) without caring about the actual validation. Since it is a bit harder with this field (having multiple formats), I'd say we can check here if the string is a "valid reference" and drop it if so, and the conditions will be replaced by:
```go
	if errs := validation.IsDNS1123Subdomain(community); len(errs) != 0 {
		communities = append(communities, community)
	}
```
(using `"k8s.io/apimachinery/pkg/util/validation"`)
this still makes the assumption that a community value can not be of this format (which k8s object names follow), but I think it's better than the specific community validations.

File: internal/config/validator.go [hunk 11]
```diff
@@ -58,13 +58,65 @@ func (v *validator) Validate(resources ...client.ObjectList) error {
...
+		}
+		clusterResources.BGPAdvs[i].Spec.Communities = communities
+		bgpAdv.Spec.Communities = communities
```
Comment (oribon) [inline internal/config/validator.go:86]: why is this needed?

File: internal/config/validator_test.go [hunk 12]
```diff
@@ -28,3 +29,255 @@ func TestValidator(t *testing.T) {
...
+		result := existCommunity(test.community, test.communities)
+		if result != test.expect {
+			t.Errorf("%s: expect is %v, but result is %v", test.community, test.expect, result)
+		}
+	}
+}
+func TestResetTransientErrorsFields(t *testing.T) {
+	tests := []struct {
+		clusterResources       ClusterResources
+		expectclusterResources ClusterResources
```
Comment (oribon) [inline internal/config/validator_test.go:89]: nit: change this to just "expected"

File: internal/config/validator_test.go [hunk 13]
```diff
@@ -28,3 +29,255 @@ func TestValidator(t *testing.T) {
...
+		}
+		for i := range result.LegacyAddressPools {
+			for j := range result.LegacyAddressPools[i].Spec.BGPAdvertisements {
+				if !compareStringArrays(result.LegacyAddressPools[i].Spec.BGPAdvertisements[j].Communities,
+					test.expectclusterResources.LegacyAddressPools[i].Spec.BGPAdvertisements[j].Communities) {
+					t.Errorf("Expect is %v, but result is %v", test.expectclusterResources.LegacyAddressPools[i].Spec.BGPAdvertisements[j].Communities,
+						result.LegacyAddressPools[i].Spec.BGPAdvertisements[j].Communities)
+				}
+			}
+		}
```
Comment (oribon) [inline internal/config/validator_test.go:231]: this can be replaced by:
```go
			if !cmp.Equal(result, expected) {
				t.Fatalf("test %s failed, unexpected clusterResources (-want +got)\n", test.desc, cmp.Diff(result, expected))
			}
```

File: internal/config/validator_test.go [hunk 14]
```diff
@@ -28,3 +29,255 @@ func TestValidator(t *testing.T) {
...
+	for _, test := range tests {
+		result := existCommunity(test.community, test.communities)
+		if result != test.expect {
+			t.Errorf("%s: expect is %v, but result is %v", test.community, test.expect, result)
+		}
+	}
+}
+func TestResetTransientErrorsFields(t *testing.T) {
+	tests := []struct {
+		clusterResources       ClusterResources
```
Comment (oribon) [inline internal/config/validator_test.go:88]: add a "desc" field

File: internal/config/validator.go (hunk 11 above)
Comment (shimritproj) [reply internal/config/validator.go:86]: Because I save the entire communities that don't generate a Transienterror, and when I finish the loop that occurs in lines 78-84, I update the 'communities' field. 
The goal is to update the field and ensure that the entire set of communities doesn't create a Transienterror.

File: internal/config/validator_test.go (hunk 14 above)
Comment (shimritproj) [reply internal/config/validator_test.go:88]: In the struct? Why? Maybe I don't understand what you mean.
Can you give me an example?

File: internal/config/validator.go (hunk 11 above)
Comment (oribon) [reply internal/config/validator.go:86]: I'm talking specifically about line 86, why isn't
```go
clusterResources.BGPAdvs[i].Spec.Communities = communities
```
enough?

File: internal/config/validator_test.go (hunk 13 above)
Comment (shimritproj) [reply internal/config/validator_test.go:231]: Thank you!! I searched for a function that does it, but I didn't find one. So, I wrote all of this. I will change it. tnx!

File: internal/config/validator_test.go (hunk 14 above)
Comment (oribon) [reply internal/config/validator_test.go:88]: look for `test.desc` in the code, together with t.Run it gives each testcase a name (description) and it makes it easier to understand which scenario failed

File: internal/config/validator.go (hunk 11 above)
Comment (shimritproj) [reply internal/config/validator.go:86]: Oh! you're right, tnx!

File: website/content/release-notes/_index.md [hunk 15]
```diff
@@ -16,6 +16,7 @@ Chores:
 - E2E: Receive prefixes using frr-k8s alongside MetalLB ([PR 2211](https://github.com/metallb/metallb/pull/2211))
 - Images updated to Go 1.20.12 ([PR 2213](https://github.com/metallb/metallb/pull/2213))
 - CI/E2E: Relabel the frr metrics from frr-k8s to show as MetalLB's ([PR 2210](https://github.com/metallb/metallb/pull/2210))
+- Webhook does not process transient errors properly ([PR 2202](https://github.com/metallb/metallb/pull/2202)), [ISSUE 2173](https://github.com/metallb/metallb/issues/2173))
```
Comment (oribon) [inline website/content/release-notes/_index.md:19]:
```suggestion
- Webhooks: avoid transient errors ([PR 2202](https://github.com/metallb/metallb/pull/2202)), [ISSUE 2173](https://github.com/metallb/metallb/issues/2173))
```

**Review by oribon — APPROVED** [review]
lgtm, thanks!
